strutil/matchcounter.go

```diff
@@ -0,0 +1,81 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package strutil
+
+import (
+	"bytes"
+	"regexp"
+)
+
+// A MatchCounter is a discarding io.Writer that retains up to N
+// matches to its Regexp before just counting matches.
+//
+// It does not work with regexps that cross newlines; in fact it will
+// probably not work if the data written isn't line-orineted.
+type MatchCounter struct {
+	// Regexp to use to find matches in the stream
+	Regexp *regexp.Regexp
+	// Maximum number of matches to keep; if < 0, keep all matches
+	N int
+
+	count   int
+	matches []string
+	partial []byte
+}
+
+func (w *MatchCounter) Write(p []byte) (int, error) {
+	n := len(p)
+	if len(w.partial) > 0 {
```

Comment:
Nitpick: would this work without the conditional check on `len(w.partial)`, I mean, the idx would just be -1.

Reply:
Can you ask that again, but slower? It's late...